onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc

```diff
@@ -1773,28 +1769,51 @@ static const HandlerInfo* GetHandler(api::NodeRef& node, bool allow_extended_ops
   return nullptr;
 }
 
-// Some op should be optimized any time there is a transpose as input and a handler is available.
-static bool CanNodeSkipCostCheck(const OptimizerCtx& ctx, const api::NodeRef& node) {
-  if (node.IsOp("Transpose")) {
-    return true;
+static int CalculateCost(const api::GraphRef& graph, const api::NodeRef& node,
+                         const std::vector<int64_t>& perm,
+                         const std::unordered_set<std::string>& outputs_leading_to_transpose,
+                         const HandlerInfo& info,
+                         const std::vector<size_t>& input_indices) {
+  // We require the input cost (number of transposes before the op) and the total cost to strictly decrease.
+  // Strict decrease of the input cost ensures the optimization is stable, since the total cost decrease is just an
+  // estimate (the transpose after the op may or may not cancel with a subsequent transpose). We don't want
+  // repeated runs of the optimizer to have a transpose toggle between two inputs of a binary op.
+  int cost = EstimateTransposeInputsCost(graph, node, perm, input_indices);
+
+  if (cost < 0 && info.transposes_outputs) {
+    // If the output will be transposed and won't ultimately cancel, factor in that cost.
+    bool has_output_leading_to_transpose = false;
+    auto outputs = node.Outputs();
+    int out_cost = 0;
+    // Having multiple outputs is rare. When it happens (Split), the total size of the outputs isn't much larger
+    // than the largest input. Cost is rank currently, so just use the largest cost (rank) over all outputs.
+    for (auto out : outputs) {
+      out_cost = std::max(out_cost, EstimateValueRank(graph, out));
+      if (outputs_leading_to_transpose.find(std::string(out)) != outputs_leading_to_transpose.end()) {
+        has_output_leading_to_transpose = true;
+      }
+    }
+
+    if (!has_output_leading_to_transpose) {
+      cost += out_cost;
+    }
   }
-  if (node.IsOp("MaxPool")) {
-    // Inclusion of MaxPool is a hack because it has higher perf in the NHWC variant when supported.
+
+  return cost;
+}
+
+// Default cost check. Returns `true` if pushing the Transpose through the node is considered to be beneficial.
+static bool DefaultCostCheck(const api::GraphRef& graph, const api::NodeRef& node,
```

Comment:
minor naming nit: it was initially surprising to me that `DefaultCostCheck` returns bool instead of CostCheckResult (i.e., it is not a CostCheckFn). and in the presence of `OrtDefaultCostCheck`, perhaps this name could be more different?

Reply:
Originally they all used bool but I had to refactor to make the CostCheckFn signature simpler. What about something like `IsBetterToPushTranspose`?

Reply:
sure, that's good